core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -860,9 +876,26 @@ private[storage] class PartiallySerializedBlock[T](
     ByteStreams.copy(unrolledBuffer.toInputStream(dispose = true), os)
     memoryStore.releaseUnrollMemoryForThisTask(memoryMode, unrollMemory)
     redirectableOutputStream.setOutputStream(os)
+
+    // Whether we use Kryo serialization
+    var kryoSerializationStream: KryoSerializationStream = null
+    if (serializationStream.isInstanceOf[KryoSerializationStream]) {
+      kryoSerializationStream = serializationStream.asInstanceOf[KryoSerializationStream]
+    }
+
     while (rest.hasNext) {
-      serializationStream.writeObject(rest.next())(classTag)
+      val value = rest.next()
+      if (kryoSerializationStream != null) {
+        if (!kryoSerializationStream.classWrote) {
+          kryoSerializationStream.writeClass(value.getClass)
```

Comment:
@srowen you can see here. Here we don't use the writeAll, because we need acquire memory according to the written size.